### PR TITLE
[1LP][RFR] Updates to container.node

### PR DESCRIPTION
**Pull request by jawatts**
Updates to container.node to further integrate with widgetastic/collectionsv3.

Updates:
Added toolbar views for Node,NodeDetails
Added accordion for NodeDetails
Updated NodeView to use BaseLoggedInPage
Removed unused impor
Commented out any import/reference to Node in existing test cases until they can be updated to use the widgetastic views/widgets 
 
Further Actions Needed:
Missing navigation to Ad hoc Metrics, External Logging, and Web Console.

### CODE REVIEW

**Comment by dajoRH** [comment]
Lint report for commit ce973c31dac6dc26245e93f6d78a1e3928b76af4:

`cfme/tests/containers/test_node.py`:
- :bangbang: Line 14:1: **E302** *expected 2 blank lines, found 1*
- :bangbang: Line 24:36: **E231** *missing whitespace after ','*
- :red_circle: Line 25:16: **F821** *undefined name 'CandidateNotFound'*
- :red_circle: Line 26:13: **F821** *undefined name 'logger'*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

File: cfme/containers/node.py [hunk 1]
```diff
@@ -120,116 +149,79 @@ def all(self):
...
-class All(CFMENavigateStep):
+class NodeAll(CFMENavigateStep):
     VIEW = NodeAllView
     prerequisite = NavigateToAttribute('appliance.server', 'LoggedIn')
 
     def step(self, *args, **kwargs):
+        """Navigate to the Nodes page"""
         self.prerequisite_view.navigation.select('Compute', 'Containers', 'Container Nodes')
 
-    def resetter(self):
```
Comment (gshefer) [inline cfme/containers/node.py]: I think the resetter is important, we assume we get List View by default for each details page.
I suggest:
```python
def resetter(self):
    from cfme.web_ui import paginator
    self.view.toolbar.view_selector.select('Grid View')
    paginator.check_all()
    paginator.uncheck_all()
```

File: cfme/containers/node.py [hunk 2]
```diff
@@ -1,91 +1,109 @@
...
+    @View.nested
+    class properties(Accordion):        # noqa
+        nav = BootstrapNav('//div[@id="ems_prop"]//ul')
+
+    @View.nested
+    class relationships(Accordion):     # noqa
+        nav = BootstrapNav('//div[@id="ems_rel"]//ul')
+
+
+class NodeDetailsEntities(View):
```
[the remark this answers is not in the document]
Comment (jawatts) [reply cfme/containers/node.py]: :+1: This is a good idea

File: cfme/containers/node.py [hunk 3]
```diff
@@ -1,91 +1,109 @@
...
+    conditions = SummaryTable(title='Conditions')
+    smart_management = SummaryTable(title='Smart Management')
+    # element attributes changed from id to class in upstream-fine+, capture both with locator
+    flash = FlashMessages('.//div[@id="flash_msg_div"]'
+                          '/div[@id="flash_text_div" or contains(@class, "flash_text_div")]')
+
+
+class NodeView(BaseLoggedInPage):
+    """A base view for all the Nodes pages"""
+    TITLE_TEXT = "Nodes"
```
[the remark this answers is not in the document]
Comment (jawatts) [reply cfme/containers/node.py]: I don't see ContainerObjectAllBaseView in the repo, so for now I will leave it how it is. I also won't be able to use the below in_node function if I am inheriting from ContainerObjectAllBaseView. Is there something equivalent in ContainerObjectAllBaseView? If not, could that be added in?

File: cfme/containers/node.py [hunk 4]
```diff
@@ -95,14 +113,25 @@ def get_random_instances(cls, provider, count=1, appliance=None):
         return [collection.instantiate(obj.name, provider)
                 for obj in itertools.islice(node_list, count)]
 
+    @property
+    def exists(self):
```
[the remark this answers is not in the document]
Comment (jawatts) [reply cfme/containers/node.py]: Since this is UI, shouldn't we check against the UI and not the backend?

File: cfme/containers/node.py (hunk 1 above)
Comment (jawatts) [reply cfme/containers/node.py]: If ressetter is necessary, I would argue to use view.entities.paginator.check_all() which is imported from BaseEntitiesView. BaseEnitiesView is conditionally set the entities based on the view selection in the UI so that you can select an entity regardless of the view. I can keep the check_all() and uncheck_all() in the resetter though

File: cfme/containers/node.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (gshefer) [reply cfme/containers/node.py]: About the ContainerObjectAllBaseView, here - https://github.com/ManageIQ/integration_tests/blob/master/cfme/containers/provider/__init__.py#L443
Let's keep this as it is right now and I'll do the following in the full widgetastic conversion PR:
Make this generic, move `def in_node()` to `ContainerObjectAllBaseView` and rename it to `def in_object()` or something, and use constant `MENU_NAME` which will be overriden in child (like `TITLE_TEXT`), i.e.:
```python
class ContainerObjectAllBaseView(ProvidersView):
...
    @property
    def in_object(self):
        """Determine if the object page is currently open"""
        return (self.logged_in_as_current_user and
                self.navigation.currently_selected == ['Compute', 'Containers', self.MENU_NAME] 
                )
    @property
    def is_displayed(self):
        """Is this page currently being displayed"""
        self.in_object and self.entities.title.text == self.TITLE_TEXT
```
and then `NodeAllView` (and actually all the rest of the objects' AllViews) could be only:
```python
class NodeAllView(ContainerObjectAllBaseView):
    TITLE_TEXT = "Nodes"
    MENU_NAME = 'Container Nodes'
```
What do you think?

File: cfme/containers/node.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (gshefer) [reply cfme/containers/node.py]: I don't think so, AFAIK you are not testing here, you are just checking if the node exists.
this is only my opinion - up to you.

File: cfme/tests/containers/test_node.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+TEST_OBJECTS = ('All', 'Details')
+
+@pytest.yield_fixture(scope="function")
+def node(provider, appliance):
+    collection = appliance.collections.nodes
+    return collection.instantiate(provider.data['hosts'][0]['name'], provider=provider)
+
+
+@pytest.mark.tier(3)
+def test_navigate_all(node):
```
[the remark this answers is not in the document]
Comment (gshefer) [reply cfme/tests/containers/test_node.py]: @jawatts I just thought maybe you can add polarion case for testing navigation and define this as product test.

File: cfme/containers/node.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jawatts) [reply cfme/containers/node.py]: :+1: This is really good. I will leave it as is for now.

File: cfme/containers/node.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jawatts) [reply cfme/containers/node.py]: I think since we are modeling the UI here, we should keep it as is. I see the same being done in a couple other places as well. Examples:
https://github.com/ManageIQ/integration_tests/blob/597ce97bed0bb329d8630f0be540235f0f17f751/cfme/cloud/tenant.py#L227
https://github.com/ManageIQ/integration_tests/blob/597ce97bed0bb329d8630f0be540235f0f17f751/cfme/cloud/stack.py#L231

File: cfme/containers/node.py (hunk 1 above)
Comment (jawatts) [reply cfme/containers/node.py]: I will add this back in

File: cfme/containers/node.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (gshefer) [reply cfme/containers/node.py]: Maybe you right. @psav what you think?

File: cfme/tests/containers/test_node.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jawatts) [reply cfme/tests/containers/test_node.py]: I can do that

File: cfme/containers/node.py [hunk 6]
```diff
@@ -1,91 +1,109 @@
-# -*- coding: utf-8 -*-
```
Comment (quarckster) [inline cfme/containers/node.py]: Why did you remove it?

File: cfme/containers/node.py [hunk 7]
```diff
@@ -1,91 +1,109 @@
...
-from cfme.exceptions import NodeNotFound
-from cfme.fixtures import pytest_selenium as sel
-from cfme.web_ui import CheckboxTable, toolbar as tb, InfoBlock, match_location
+from wrapanapi.containers.node import Node as ApiNode
+
+from cfme.base.login import BaseLoggedInPage
+from cfme.common.vm_views import ManagePoliciesView
+from cfme.exceptions import ItemNotFound, InstanceNotFound
+from cfme.common import WidgetasticTaggable, TagPageView
+from cfme.containers.provider import ContainersProvider, Labelable
```
Comment (quarckster) [inline cfme/containers/node.py]: **REQUIRED**
Please sort imports alphabetically

File: cfme/tests/containers/test_breadcrumbs.py [hunk 8]
```diff
@@ -12,7 +12,7 @@
 from cfme.containers.pod import Pod
 from cfme.containers.image_registry import ImageRegistry
 from cfme.containers.image import Image
-from cfme.containers.node import Node
+# from cfme.containers.node import Node
```
Comment (quarckster) [inline cfme/tests/containers/test_breadcrumbs.py]: **REQUIRED**
Add TODO here and further

File: cfme/tests/containers/test_containers_summary.py [hunk 9]
```diff
@@ -22,7 +22,9 @@
     pytest.mark.tier(1)]
 pytest_generate_tests = testgen.generate([ContainersProvider], scope='function')
 
-tested_objects = [Service, Route, Project, Pod, Image, Container, ImageRegistry, Node]
+# TODO Update test to use new Node views and widgets
+tested_objects = [Service, Route, Project, Pod, Image, Container, ImageRegistry  # , Node
+                  ]
```
Comment (quarckster) [inline cfme/tests/containers/test_containers_summary.py]: **REQUIRED**
It looks ugly, please add `Node` into TODO section

File: cfme/containers/node.py (hunk 6 above)
Comment (jawatts) [reply cfme/containers/node.py]: Added it back in

File: cfme/containers/node.py (hunk 7 above)
Comment (jawatts) [reply cfme/containers/node.py]: :+1:

File: cfme/tests/containers/test_containers_summary.py (hunk 9 above)
Comment (jawatts) [reply cfme/tests/containers/test_containers_summary.py]: :+1:

File: cfme/tests/containers/test_breadcrumbs.py (hunk 8 above)
Comment (jawatts) [reply cfme/tests/containers/test_breadcrumbs.py]: :+1: I went ahead and just removed import. Can be added back in when is updated

**Comment by jawatts** [comment]
@quarckster Addressed all of your comments.

File: cfme/containers/node.py [hunk 10]
```diff
@@ -1,91 +1,111 @@
...
-        return (
-            self.logged_in_as_current_user and
-            self.navigation.currently_selected == ['Compute', 'Containers', 'Container Nodes'] and
-            match_page()  # No summary, just match controller and title
-        )
+    def in_node(self):
+        """Determine if the Nodes page is currently open"""
+        return (self.logged_in_as_current_user and
+                self.navigation.currently_selected == ['Compute', 'Containers', 'Container Nodes']
+                )
```
Comment (quarckster) [inline cfme/containers/node.py]: **REQUIRED**
```python
return (
    self.logged_in_as_current_user and
    self.navigation.currently_selected == ['Compute', 'Containers', 'Container Nodes']
)

File: cfme/containers/node.py [hunk 11]
```diff
@@ -1,91 +1,111 @@
...
+    including_entities = View.include(BaseEntitiesView, use_parent=True)
+
     @property
     def is_displayed(self):
+        """Is this page currently being displayed"""
         return (
-            self.in_cloud_instance and
-            match_page(summary='Nodes')
+            self.in_node and self.entities.title.text == 'Nodes'
         )
```
Comment (quarckster) [inline cfme/containers/node.py]: **REQUIRED**
```python
return self.in_node and self.entities.title.text == 'Nodes'
```

File: cfme/containers/node.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (psav) [reply cfme/containers/node.py]: ```.exists``` here should check the UI
Methods against the models are related to the objects in MiQ therefore it does not make sense to check against the mgmt system. There are many places in the codebase right now where ```self.exists``` on an object checks either via REST or UI, but always MiQ. That's kinda why ```self.mgmt.exists``` is there, so that you can do a check of the management system.

File: cfme/containers/node.py (hunk 10 above)
Comment (jawatts) [reply cfme/containers/node.py]: :+1:

**Comment by jawatts** [comment]
@quarckster Fixed the formatting issues. Thanks for the review!

File: cfme/tests/containers/test_node.py [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+@pytest.mark.polarion('CMP-10782')
+@pytest.mark.tier(3)
+@pytest.mark.provider([ContainersProvider])
+def test_navigate_all(node):
+    for test_item in TEST_VIEWS:
+        try:
+            view = navigate_to(node, test_item)
+        except CandidateNotFound:
+            logger.warning(
+                'Exception while attempting to navigate to page')
```
Comment (mfalesni) [inline cfme/tests/containers/test_node.py]: **REQUIRED/QUESTION**: Shouldn't this be a test failure as well? Eg. using soft_assert?

Comment (jawatts) [reply cfme/tests/containers/test_node.py]: @mfalesni Yes, it should be. Let me fix that.

Comment (jawatts) [reply cfme/tests/containers/test_node.py]: :+1: done

**Comment by psav** [comment]
Moving to WIP for answer to review question

**Comment by jawatts** [comment]
@psav This is some technical debt we will have until @gshefer is done the widegastic conversion. We will have to rewrite our tests to remove web_ui. I used iPython to test navigating to the Container Nodes Page, sorting the table, checking/unchecking boxes in the table, navigating to the details of a particular Node, and selecting the toolbar widgets.

File: cfme/containers/node.py [hunk 13]
```diff
@@ -238,28 +238,31 @@ class Utilization(CFMENavigateStep):
...
-        self.prerequisite_view.monitor.item_select('Utilization')
+        """Navigate to the Utilization page"""
+        self.prerequisite_view.toolbar.monitoring.item_select('Utilization')
 
 
-class NodeTimelinesForm(NodeView):
-    # TODO PR 3710, timeline widget
+class NodeTimelinesView(NodeView, TimelinesView):
+    """Timeline page for Nodes"""
+    VIEW = TimelinesView
```
Comment (mfalesni) [inline cfme/containers/node.py]: **REQUIRED**: Use the `Ignore` decorator here to avoid the TimelinesView being instantiated automatically if you intend this to be a constant. Though this looks to me like some sort of an artifact copied over from a step.

**Comment by gshefer** [comment]
once @mfalesni comment addressed - LGTM :+1:

File: cfme/containers/node.py (hunk 13 above)
Comment (jawatts) [reply cfme/containers/node.py]: :+1: done
